### PR TITLE
[SPARK-36428][SQL] the seconds parameter of make_timestamp should accept integer type

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
With ANSI mode, `SELECT make_timestamp(1, 1, 1, 1, 1, 1)` fails, because the 'seconds' parameter needs to be of type DECIMAL(8,6), and INT can't be implicitly casted to DECIMAL(8,6) under ANSI mode.

```
org.apache.spark.sql.AnalysisException
cannot resolve 'make_timestamp(1, 1, 1, 1, 1, 1)' due to data type mismatch: argument 6 requires decimal(8,6) type, however, '1' is of int type.; line 1 pos 7
```

We should update the function `make_timestamp` to allow integer type 'seconds' parameter.

### Why are the changes needed?
Make `make_timestamp` could accepts integer as 'seconds' parameter.

### Does this PR introduce _any_ user-facing change?
'Yes'.
`make_timestamp` could accepts integer as 'seconds' parameter.

### How was this patch tested?
New tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -2507,12 +2507,15 @@ case class MakeTimestampLTZ(
...
        2019-07-01 00:00:00
+      > SELECT _FUNC_(2019, 6, 30, 23, 59, 1);
+       2019-06-30 23:59:01
       > SELECT _FUNC_(2019, 13, 1, 10, 11, 12, 'PST');
        NULL
       > SELECT _FUNC_(null, 7, 22, 15, 30, 0);
        NULL
   """,
   group = "datetime_funcs",
+  note = "",
```
Comment (beliefer) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2518]: I will add some comments here after some discussion.

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
retest this please

**Comment by beliefer** [comment]
ping @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 2]
```diff
@@ -2507,12 +2507,17 @@ case class MakeTimestampLTZ(
...
        2019-07-01 00:00:00
+      > SELECT _FUNC_(2019, 6, 30, 23, 59, 1);
+       2019-06-30 23:59:01
       > SELECT _FUNC_(2019, 13, 1, 10, 11, 12, 'PST');
        NULL
       > SELECT _FUNC_(null, 7, 22, 15, 30, 0);
        NULL
   """,
   group = "datetime_funcs",
+  note = """
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2518]: instead of adding a note, shall we just update the parameter doc of `sec` and say it can be either decimal(8, 6) or int?

File: sql/core/src/test/resources/sql-tests/inputs/timestamp.sql [hunk 3]
```diff
@@ -18,6 +18,9 @@ select localtimestamp() = localtimestamp();
 SELECT make_timestamp(2021, 07, 11, 6, 30, 45.678);
 SELECT make_timestamp(2021, 07, 11, 6, 30, 45.678, 'CET');
 SELECT make_timestamp(2021, 07, 11, 6, 30, 60.007);
+SELECT make_timestamp(1, 1, 1, 1, 1, 1);
+SELECT make_timestamp(1, 1, 1, 1, 1, 60);
+SELECT make_timestamp(1, 1, 1, 1, 1, 61);
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/inputs/timestamp.sql:23]: can we also check `make_timestamp(1, 1, 1, 1, 1, null)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 2 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2518]: OK

**Comment by cloud-fan** [comment]
@beliefer can you fix the code conflicts? thanks!

**Comment by beliefer** [comment]
> @beliefer can you fix the code conflicts? thanks!

OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 4]
```diff
@@ -2495,7 +2495,9 @@ case class MakeTimestampLTZ(
       * hour - the hour-of-day to represent, from 0 to 23
       * min - the minute-of-hour to represent, from 0 to 59
       * sec - the second-of-minute and its micro-fraction to represent, from
-              0 to 60. If the sec argument equals to 60, the seconds field is set
+              0 to 60. The sec argument can be either decimal(8, 6) or int.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2498]: maybe make it more user-facing.
```
... 0 to 60. The value can either be an integer like 13, or a fraction like 13.123
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 5]
```diff
@@ -2617,30 +2628,41 @@ case class MakeTimestamp(
...
     val d = Decimal.getClass.getName.stripSuffix("$")
+    val decimalValue = ctx.freshName("decimalValue")
     val failOnErrorBranch = if (failOnError) "throw e;" else s"${ev.isNull} = true;"
     nullSafeCodeGen(ctx, ev, (year, month, day, hour, min, secAndNanos, timezone) => {
       val zoneId = timezone.map(tz => s"$dtu.getZoneId(${tz}.toString())").getOrElse(zid)
       val toMicrosCode = if (dataType == TimestampType) {
         s"""
            |java.time.Instant instant = ldt.atZone($zoneId).toInstant();
            |${ev.value} = $dtu.instantToMicros(instant);
-           |""".stripMargin
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2635]: unnecessary change. Actually the previous code style is auto-generated by IDEA.

File: sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out [hunk 6]
```diff
@@ -107,6 +107,40 @@ java.time.DateTimeException
...
+Invalid value for SecondOfMinute (valid values 0 - 59): 61
+
+
+-- !query
+SELECT make_timestamp(1, 1, 1, 1, 1, null)
+-- !query schema
+struct<>
+-- !query output
+org.apache.spark.sql.AnalysisException
+cannot resolve 'make_timestamp(1, 1, 1, 1, 1, NULL)' due to data type mismatch: argument 6 requires (decimal(8,6) or int) type, however, 'NULL' is of void type.; line 1 pos 7
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out:141]: What's the previous behavior? also fail?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 4 above)
Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2498]: OK

File: sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out (hunk 6 above)
Comment (beliefer) [reply sql/core/src/test/resources/sql-tests/results/ansi/timestamp.sql.out:141]: Let's add NullType support.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 7]
```diff
@@ -2494,8 +2494,10 @@ case class MakeTimestampLTZ(
       * day - the day-of-month to represent, from 1 to 31
       * hour - the hour-of-day to represent, from 0 to 23
       * min - the minute-of-hour to represent, from 0 to 59
-      * sec - the second-of-minute and its micro-fraction to represent, from
-              0 to 60. If the sec argument equals to 60, the seconds field is set
+      * sec - the second-of-minute and its micro-fraction to represent, from 0 to 60.
+              The value can be either an integer like 13 , or a fraction like 13.123.
+              If the specified sec is an int, this value represents seconds.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2499]: We can remove this now. It's clear enough that this parameter means seconds

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2499]: OK

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 8]
```diff
@@ -2556,13 +2560,20 @@ case class MakeTimestamp(
   // Accept `sec` as DecimalType to avoid loosing precision of microseconds while converting
   // them to the fractional part of `sec`.
   override def inputTypes: Seq[AbstractDataType] =
-    Seq(IntegerType, IntegerType, IntegerType, IntegerType, IntegerType, DecimalType(8, 6)) ++
-    timezone.map(_ => StringType)
+    Seq(IntegerType, IntegerType, IntegerType, IntegerType, IntegerType,
+      TypeCollection(DecimalType(8, 6), IntegerType)) ++ timezone.map(_ => StringType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2564]: where do you add the null type?

Comment (beliefer) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2564]: The conflict is not properly resolved.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2 (fixes a problem in ANSI mode)!

**Comment by beliefer** [comment]
@cloud-fan Thank you for your work

**Comment by gengliangwang** [comment]
I think we can change the accept type as `decimal(12,2)` instead, which is wider than the Integer type. So that we can get rid of the type collection and make the implementation simpler.

**Comment by beliefer** [comment]
> I think we can change the accept type as `decimal(16,6)` instead, which is wider than the Integer type. So that we can get rid of the type collection and make the implementation simpler.

Do you know the reason why the previous code uses `decimal(8,6)` ?

**Comment by gengliangwang** [comment]
I think it is trying to limit the decimal within 60.0. But we have to do a runtime check anyway, the input can be 99.999999
